File: apps/mobile/src/domain/scheduler.ts
// Repetição espaçada por ITEM (docs/adr/0003). Funções puras com `now` injetado.
// A variável central é a meia-vida: é exatamente o que o ranking no servidor vai prever depois, então o
// armazenamento não muda quando o modelo entrar — só a fórmula de `halfLifeHours` é substituída.
export type ItemState = {
  itemId: string;
  unitId: string;
  /** Cresce 1 por acerto, cai 2 por erro, nunca negativo. */
  strength: number;
  halfLifeHours: number;
  dueAt: string;
  reps: number;
  lapses: number;
  consecutiveCorrect: number;
  lastResult: boolean;
  lastSeenAt: string;
  updatedAt: string;
};

export const BASE_HOURS = 4;
export const MIN_HALF_LIFE_HOURS = 4;
export const MAX_HALF_LIFE_HOURS = 90 * 24;

export const halfLifeHours = (strength: number): number => Math.min(MAX_HALF_LIFE_HOURS, Math.max(MIN_HALF_LIFE_HOURS, BASE_HOURS * 2 ** strength));

export function applyResult(current: ItemState | undefined, input: { itemId: string; unitId: string; correct: boolean; now: Date }): ItemState {
  const strength = input.correct ? (current?.strength ?? 0) + 1 : Math.max(0, (current?.strength ?? 0) - 2);
  const halfLife = halfLifeHours(strength);
  const seenAt = input.now.toISOString();
  return {
    itemId: input.itemId, unitId: input.unitId, strength, halfLifeHours: halfLife,
    dueAt: new Date(input.now.getTime() + halfLife * 60 * 60 * 1000).toISOString(),
    reps: (current?.reps ?? 0) + 1, lapses: (current?.lapses ?? 0) + (input.correct ? 0 : 1),
    consecutiveCorrect: input.correct ? (current?.consecutiveCorrect ?? 0) + 1 : 0,
    lastResult: input.correct, lastSeenAt: seenAt, updatedAt: seenAt,
  };
}

export const isDue = (state: ItemState, now: Date): boolean => Date.parse(state.dueAt) <= now.getTime();
/** Já errou e ainda não recuperou três acertos seguidos. */
export const isWeak = (state: ItemState): boolean => state.lapses > 0 && state.consecutiveCorrect < 3;
/** Conta para a conclusão da unidade: o último encontro foi um acerto. */
export const isMastered = (state: ItemState): boolean => state.consecutiveCorrect >= 1;

File: apps/api/app/ranking/scheduler.py
"""Porta em Python de apps/mobile/src/domain/scheduler.ts, usada pelo replay para reconstruir o estado antes de cada tentativa."""
from dataclasses import dataclass, replace
from datetime import datetime, timedelta

BASE_HOURS = 4.0
MIN_HALF_LIFE_HOURS = 4.0
MAX_HALF_LIFE_HOURS = 90.0 * 24


def half_life_hours(strength: int) -> float:
    return min(MAX_HALF_LIFE_HOURS, max(MIN_HALF_LIFE_HOURS, BASE_HOURS * 2 ** strength))


@dataclass(frozen=True)
class State:
    item_id: str
    unit_id: str | None
    strength: int
    half_life_hours: float
    due_at: datetime
    reps: int
    lapses: int
    consecutive_correct: int
    last_result: bool
    last_seen_at: datetime


def apply_result(current: State | None, *, item_id: str, unit_id: str | None, correct: bool, now: datetime) -> State:
    strength = (current.strength if current else 0) + 1 if correct else max(0, (current.strength if current else 0) - 2)
    half_life = half_life_hours(strength)
    return State(
        item_id=item_id, unit_id=unit_id, strength=strength, half_life_hours=half_life, due_at=now + timedelta(hours=half_life),
        reps=(current.reps if current else 0) + 1, lapses=(current.lapses if current else 0) + (0 if correct else 1),
        consecutive_correct=(current.consecutive_correct if current else 0) + 1 if correct else 0,
        last_result=correct, last_seen_at=now,
    )


def is_due(state: State, now: datetime) -> bool:
    return state.due_at <= now


def is_weak(state: State) -> bool:
    return state.lapses > 0 and state.consecutive_correct < 3


def is_mastered(state: State) -> bool:
    return state.consecutive_correct >= 1


def with_time(state: State, **fields) -> State:
    return replace(state, **fields)
